File: day012_greedy_conversions/related_jump_game/solution.py
import sys
from typing import List

class Solution:
    # Optimized: O(N) Time, O(1) Space - Backward Greedy Traversal
    def canJump_optimal(self, nums: List[int]) -> bool:
        if not nums: return False
        
        goal = len(nums) - 1 # The goal is the last index

        # Iterate backward from the second-to-last element
        for i in range(len(nums) - 2, -1, -1):
            # Check if current position can reach or pass the goal
            if i + nums[i] >= goal:
                goal = i # This is the new, closer goal
                
        # If the goal successfully reached index 0, return True
        return goal == 0

    # Brute Force: O(N^2) Time - Dynamic Programming (DP) or recursion with memoization
    def canJump_brute(self, nums: List[int]) -> bool:
        # Simplistic O(N^2) DP solution (often the first attempt)
        N = len(nums)
        memo = [False] * N
        memo[N - 1] = True # Last index is reachable

        for i in range(N - 2, -1, -1):
            max_jump = min(i + nums[i], N - 1)
            for j in range(i + 1, max_jump + 1):
                if memo[j]:
                    memo[i] = True
                    break
        
        return memo[0]

# Standard input reading (assuming the input is a single line of space-separated jumps)
def solve_jump_game():
    try:
        # Read the array elements
        nums = list(map(int, sys.stdin.readline().split()))
    except:
        return
    
    result = Solution().canJump_optimal(nums)
    print("YES" if result else "NO")

if __name__ == "__main__":
    solve_jump_game()

File: day012_greedy_conversions/related_jump_game/solution.java

class Solution {

    // Optimized: O(N) Time, O(1) Space - Backward Greedy Traversal
    public boolean canJump_optimal(int[] nums) {
        if (nums == null || nums.length == 0) return false;
        
        int goal = nums.length - 1;

        // Iterate backward
        for (int i = nums.length - 2; i >= 0; i--) {
            // Check if current position can reach or pass the goal
            if (i + nums[i] >= goal) {
                goal = i; // Update goal
            }
        }
        
        return goal == 0;
    }

    // Brute Force: O(N^2) Time - Dynamic Programming
    public boolean canJump_brute(int[] nums) {
        int N = nums.length;
        if (N <= 1) return true;
        
        boolean[] memo = new boolean[N];
        memo[N - 1] = true;

        for (int i = N - 2; i >= 0; i--) {
            int maxJump = Math.min(i + nums[i], N - 1);
            for (int j = i + 1; j <= maxJump; j++) {
                if (memo[j]) {
                    memo[i] = true;
                    break;
                }
            }
        }
        return memo[0];
    }
}